MVR: Skeleton for new router structure

scgruber opened the pull request:
This commit defines the basic structure for making a router using the
new framework with tokio 0.2.

In main() we create:
  - ingressors
  - egressors
  - tokio runtime
  - Route instance

We then pass these to run_route() to wire everything together and
initiate the tokio threads. run_route() should be moved into
route-rs-runtime before all of this is actually merged, since it's
pretty generic over any route link.

Since we don't have af_packet integrated with links yet, we're still
setting up InputChannelLink and OutputChannelLink at either end of
the router. This new framework does handle multiple inputs and output
channels, though.

While this does compile, it probably won't run because a lot of things
are left empty. I'll fill them in with reasonable defaults before this
is merged. A following commit will implement the top-level internals
of the minimum viable router, and then a further commit will convert
graphgen and the other examples into this style.

scgruber committed:
MVR: Skeleton for new router structure

This commit defines the basic structure for making a router using the
new framework with tokio 0.2.

In main() we create:
  - ingressors
  - egressors
  - tokio runtime
  - Route instance

We then pass these to run_route() to wire everything together and
initiate the tokio threads. run_route() should be moved into
route-rs-runtime before all of this is actually merged, since it's
pretty generic over any route link.

Since we don't have af_packet integrated with links yet, we're still
setting up InputChannelLink and OutputChannelLink at either end of
the router. This new framework does handle multiple inputs and output
channels, though.

While this does compile, it probably won't run because a lot of things
are left empty. I'll fill them in with reasonable defaults before this
is merged. A following commit will implement the top-level internals
of the minimum viable router, and then a further commit will convert
graphgen and the other examples into this style.

CollinValley commented on examples/minimum-viable-router/src/main.rs:
```diff
@@ -0,0 +1,62 @@
+    runtime: &mut tokio::runtime::Runtime,
+    route: route::Route,
+) {
+    let mut all_runnables = vec![];
+
+    let (_, upstreams): (
+        Vec<Vec<TokioRunnable>>,
+        Vec<Vec<PacketStream<EthernetFrame>>>,
+    ) = ingressors
+        .into_iter()
+        .map(|receiver| InputChannelLink::new().channel(receiver).build_link())
+        .unzip();
+
+    let (mut route_runnables, downstreams) = route
+        .ingressors(upstreams.into_iter().flatten().collect())
+        .build_link();
+
+    all_runnables.append(&mut route_runnables);
+
+    let egressor_inputs = downstreams.into_iter().zip(egressors.into_iter());
+
+    let (egress_runnables, _): (Vec<Vec<TokioRunnable>>, Vec<Vec<PacketStream<_>>>) =
+        egressor_inputs
+            .map(|(ingressor, sender)| {
+                OutputChannelLink::new()
+                    .ingressor(ingressor)
+                    .channel(sender)
+                    .build_link()
+            })
+            .unzip();
+
+    all_runnables.append(&mut egress_runnables.into_iter().flatten().collect());
+
```
All this building needs to happen inside the block_on block, otherwise you can get initialization failures.  The block on provides the runtime context. It will be a rare error, but if people have external timer links and stuff it won't run.

CollinValley replied:
Especially because if Nathans links are initialized outside of the "router", those will need the tokio context active since they need to access the reactor and executor threads of tokio.

scgruber replied:
The way I envisioned this is that the `ingressors` and `egressors` would be replaced with the socket objects and we'd still assemble the links inside of `run_route`. Probably given that we should have a less generic name for this method, since it bakes in the use of `af_packet`.

But it should be relatively okay to just extend the `block_on` around the rest of this method body.

CollinValley replied:
yeah, that makes sense to me. The block_on thing is annoying,  Another annoyance is that the `route` object needs to be also built inside a tokio context.  So we should probably change that argument to something like `route: Fn() -> Link`, then we can call `let router = route::new();` inside the tokio `block_on`.  For an example of this, I did it inside the prototype run harness in the runtime, under `utils`

CollinValley replied:
The Fn doesn't have to return a `Link` either, could be some other struct, but returning the `Link` makes some sense from our composability story.

scgruber replied:
Do we have to initialize the `route` object fully inside the `block_on`? Or can we just `build_link()` in there. Keeping it outside allows the implementor to pass additional setup fields in `main()` if desired.

CollinValley replied:
Only the "Link" objects need to be initialized inside the block_on, so yes we could pre-init all the builder syntax stuff outside the context, then pass the object that you are about to build_link() on into the context, and call build_link.   I think that is a pretty good idea.

scgruber committed:
MVR: Implement interface encap/decap

This handles tagging raw ethernet frames with their source and
destination interfaces. This will allow us to make routing decisions
based on where things come from and declare where they are going.

A future commit will implement the processors for actually modifying
the tags, for now we just set them at the beginning and throw them
away.

Note that the current router just reflects all traffic back to its
origin interface. This would obviously be bad, but is done for now
to break the implementation apart into separate pieces.

scgruber committed:
MVR: Join and dispatch by Interfaces

This takes another step towards the middle of the router, unifying the
input streams into one so we can add the packet processing pipeline.

One interesting thing about this is that we need to implement
BlackHoleLink after all, because something needs to poll for handling
packets that have somehow gotten through without any tagging. It would
be nice to be able to prove that we can't get None at the end of this
using the type system, but I'm not sure how to do it without requiring
that we make all egress port routing decisions in the same place, which
seems like a really annoying limitation. I have therefore opted to
black hole any packets which make it all the way through in this state,
because crashing the router on them feels too harsh.

scgruber committed:
MVR: Add tracers after join and before dispatch

This will allow us to see packets entering and exiting the router.

Currently the Trace processor doesn't do anything; it's just a copy of
the Identity processor.

scgruber committed:
MVR: Stub out protocol classifier

This will allow us to split into different processing chains for the
protocols that we are supporting in this router. All other types of
packets are dropped because we don't know what to do with them.

Note also that the classification logic is unimplemented for now. We
just treat everything as unknown, so all data will be dropped.

scgruber committed:
MVR: Handling for the five supported protocols

Since we're leaving it to individual implementors how the protocol
handling is a done, here they are just represented by a ProcessLink
with an identity element. This should be sufficient to get everybody
out of each other's way during the implementation phase.

CollinValley commented on examples/minimum-viable-router/src/main.rs:
```diff
@@ -0,0 +1,85 @@
+    let egress_lan = crossbeam::channel::unbounded();
+    let egress_host = crossbeam::channel::unbounded();
+
+    drop(ingress_wan.0);
```
Why are we doing this? We can just let Rust do the dropping for us? Also won't the recv close as soon as we do this?

scgruber replied:
That is the point. It looks a little silly here, but if we actually had packets to pass through, this would allow us to run a test in the `main()`. If we don't explicitly drop the channels, `run_route()` will never exit because nothing stops the router.

ddowl commented on route-rs-runtime/src/link/primitive/black_hole_link.rs:
```diff
@@ -0,0 +1,91 @@
+use crate::link::{Link, LinkBuilder, PacketStream};
```
I thought we got rid of the black hole links in favor of moving that logic to Processors?

scgruber replied:
There's no way to poll the links, so we'd have to do something really strange like join the drop path back into one of the actual flows of the router.

CollinValley commented on examples/minimum-viable-router/src/main.rs:
```diff
@@ -0,0 +1,85 @@
+    println!("Congratulations! Your router successfully does nothing!")
+}
+
+// TODO: Move this to route-rs-runtime
```
This looks exactly like what already exists in the runtime under the utils called "runner"

scgruber replied:
`runner` doesn't provide a way to configure the inputs to the overall `Link`. I suppose you could put it all into the thunk that returns the `Link`, but I'm trying to keep that logic inside the library.

Once we have `af_packet`, this method will handle setting up all of those links and take the socket paths instead of channels.

scgruber replied:
May be able to refactor this thing to call `runner` though.

CollinValley replied:
Hmm, I just want runner to be the thing that we use.  I'm not married to the current implementation of runner per se, just that this commit stack should make runner fit our purpose here. Let me think on it for a few days and find a way to merge everything. This use case is valid.
